Add A4 Pitch column (schema + UI label)

quaklet opened the pull request:
Continuing work on https://github.com/mixxxdj/mixxx/pull/15729

Rebased @Pappmann's commits on current [`main`](https://github.com/mixxxdj/mixxx/tree/main).
Implemented the remaining change requests from @daschuer.

Here's a screenshot of what the column looks like – for the purpose of the screenshot I assigned Hz values randomly.

<img width="84" height="352" alt="tuning frequency symbol" src="https://github.com/user-attachments/assets/78a0ba59-d382-4107-8270-f7cba96fa3ed" />

I'm not sure what to do next.

Currently the frequency info is stored in the SQLite DB both as part of the `keys` proto buffer as well as as separate `tuning_frequency_hz` field. I don't know if there's a reason for doubling this information.
Are there cases where it makes sense that these two values would differ from each other?
Should the information be removed from one of the places, so either remove it from the protobuf, or drop the `tuning_frequency_hz` column from the `library` table?

From https://github.com/mixxxdj/mixxx/pull/15729#discussion_r2611304975:

> In this case here I am happy to merge a very basic tuning column version quickly and iterate from there. My idea is to adjust my Analyzer prototype branch for using this column and than you are in the reviewer role.
> 
> The core problem I like to fix with all these pitch stuff is this: People us the color wheel to find compatible tracks, but visual compatible tracks can be totally out of tune even worse than a less compatible track in tune. My idea was to mark these tracks as "requires pitching". Your arrow idea is well suited.

> I have probably not understand your 440 Hz / 432 Hz side by side idea. What makes a 432 Hz tracks special in a 440 Hz set except that it is not compatible? In an 432 Hz set, native 432 Hz track are of cause first class citizen with no ↓ or ↑.
> 
> Would an either/or solution work for your with tuning preferences option?

So should more work be done on how the data is displayed in Mixxx, or what's next?
What's blocking a merge?

Commit:
Add tuning frequency persistence and library column

Commit:
Ensure tuning_frequency column is auto-created

Commit:
Rename tuning column to A4 Pitch

Commit:
Fix build: include QSqlError in mixxxdb.cpp

Commit:
Apply PR feedback: Change to double/FLOAT, remove is_432hz

- Change database schema from INTEGER to FLOAT for cents precision
- Change protobuf tuning_frequency_hz from int32 to double
- Remove is_432hz field from protobuf (redundant)
- Remove is432Hz() methods from Keys and Track classes
- Update all tuning frequency handling to use double
- Use QStringLiteral for unicode symbols in keydelegate
- Add 0.5 Hz tolerance for 432Hz detection
- Remove k432HzRole, use only kTuningFrequencyRole

Addresses feedback from PR #15729 review by @daschuer

Commit:
Fix: Use escaped Unicode per coding guidelines

Use \uXXXX notation instead of literal Unicode characters
as per Mixxx coding guidelines for better portability.

Commit:
Apply PR feedback: Remove redundant checks

- Remove redundant DB state checks

Addresses feedback from PR #15729 review by @daschuer

Commit:
fix: use `double` for Hz values

Fixes a compiler warning about implicit type conversion.

Commit:
fix: avoid gap in protobuf field numbers

daschuer commented:
Thank you for jumping in. 

> Currently the frequency info is stored in the SQLite DB both as part of the keys proto buffer as well as as separate tuning_frequency_hz field.

I am unsure either. The issue is that this information in the protobuf makes its complete. But having it in the library is also required for fast access. in connection with the key column. So the current situation is kind of OK. 

I stumbled over another issue. This PR assumes that all existing entries are 440 Hz. This is OK for the use aspects inside Mixxx. 
However we can later no longer distinguish if this is a confirmed 440Hz track (by user or analyser) or just a defaulted value that can be wrong. 
Can we use 0 or such as default for data storage and treat it as 440 Hz in Mixxx?

daschuer commented:
The next step would be an edit field in the track preferences and extending the Analyzer to detect the tuning. I have a working prototype for the later. I will issue a PR for it based on your one here.

Commit:
fix: use an identifiable default frequency for pitch

Use 0 as default for the pitch frequency in the database, to allow to
identify those library entries that are using a default value, instead
of the detected frequency.

Commit:
fix: use 10 cents tolerance for tuning frequency symbols

We assume 10 cents to be the lower bound for an audible difference in
key for humans.

quaklet commented on src/library/tabledelegates/keydelegate.cpp:
```diff
@@ -1,11 +1,24 @@
+constexpr int kTuningSymbolWidth = 14;
+constexpr double kStandardTuningHz = 440.0;
+constexpr double k432Hz = 432.0;
+constexpr double k432HzTolerance = 0.5; // Allow 0.5 Hz tolerance for 432Hz detection
```
Done. ✅ 

I did some quick research and 10 cents looks good to me, too.
For example [here on Wikipedia they cite references which give 5-6 cents](https://en.wikipedia.org/wiki/Cent_(music)#Human_perception), and also some [example audio files to test your own perception](https://en.wikipedia.org/wiki/Cent_(music)#Sound_files).
To check the 2.5 Hz value I used this function and came to about the same value
```py
>>> f = lambda hz, ct: hz * 2 ** (ct/1200) - hz
>>> f(432, 10)
2.5025505413127007
>>> f(440, 10)
2.548894069855521
```

Sadly I think we can't use a function like that directly in our code (without runtime cost), because [`std::pow` isn't `constexpr` until C++26](https://en.cppreference.com/w/cpp/numeric/math/pow); but a fixed value of 2.5 seems good enough. :)
```cpp
constexpr double offsetHz(double refHz, int offsetCents) noexcept {
    return refHz * std::pow(2.0, offsetCents / 1200.0);
}
```

<img width="174" height="528" alt="10cent" src="https://github.com/user-attachments/assets/885cba41-7a52-4be6-94d8-7bfc9e470ca0" />

daschuer commented on src/track/track.cpp:
```diff
@@ -1759,6 +1776,29 @@ ExportTrackMetadataResult Track::exportMetadata(
+                const int centsRounded = static_cast<int>(std::lround(cents));
+                if (centsRounded != 0) {
+                    const QString offsetText = centsRounded > 0
+                            ? QStringLiteral("+%1").arg(centsRounded)
```
The tuning info is not standard. There might be other application that refuse reading the key. So we may want to make this a preferences option. But we can also just do it like RapidEvolution and wait for users requesting this feature.

daschuer commented on src/track/track.cpp:
```diff
@@ -1759,6 +1776,29 @@ ExportTrackMetadataResult Track::exportMetadata(
+        // Encode tuning offset (RapidEvolution style) into key text for tag roundtrip.
+        // Keep the database value untouched; only the exported metadata is modified.
+        const double tuningHz = m_record.getKeys().getDetectedTuningFrequencyHz();
+        if (tuningHz > 0.0 && tuningHz != kDefaultTuningFrequency) {
```
```suggestion
        if (tuningHz > 0.0) {
```
You are right, this can be removed. This will also exort "Am +00" as an indicator that we are sure it is a 440 Hz Am. "Am" alone means it is A minor, but I don't know at which tuning.

quaklet replied:
I prepared a commit for this. If I understand you correctly, the export should always use a string formatted something like "{key} {cents:+03}", ie. "Am +00", "B -30", "C +23", etc.
When wanting to test this I noticed that the code is apparently never called. I tried to find out when it should be called, but it's a bit confusing to me at the moment. Can you give me a hint what operation in Mixxx should cause this code to be executed – like what UI elements do I need to click on, or what else I need to do to properly test this?
I found a test (`TrackMetadataExportTest.keepWithespaceKey`) that calls the function, but there the `ImportResult` returned from `metadataSource.importTrackMetadataAndCoverImage` is always `Unavailable`, instead of `Succeeded`; so again, the code is never executed.

Commit:
fix: always include cents in the tuning info export

daschuer commented on src/proto/keys.proto:
```diff
@@ -49,4 +49,12 @@ message KeyMap {
+  // Stored as double to preserve cents precision during roundtrips
+  // This allows calculating the exact pitch offset needed for tuning adjustment:
+  // offset_semitones = log2(target_hz / tuning_frequency_hz) * 12
+  optional double tuning_frequency_hz = 5 [ default = 0.0 ];
```
I just realized that the every key info has also a tuning frequency. I have first assumed that it is constant for tracks, because artist will not tuner their instrument during a track. But we we can have DJ Mixes with different tracks and different tunings. 

Can you rename that to `global_tuning_frequency_hz`  and add a `tuning_frequency_hz field to `message KeyChange`

daschuer committed:
Speed up kTuningFrequencyRole look up

daschuer commented on src/library/basetracktablemodel.cpp:
```diff
@@ -461,6 +461,18 @@ QVariant BaseTrackTableModel::data(
+                return pTrack->getTuningFrequencyHz();
+            }
+        }
+        return Track::kDefaultTuningFrequency;
```
creating a Track object is expensive and should not be done when scrolling through the library. This is the better alternative: 
```suggestion
            return rawSiblingValue(
                    index,
                    ColumnCache::COLUMN_LIBRARYTABLE_TUNING_FREQUENCY);
        }
        return QVariant(); // unused
```

daschuer commented on src/library/basetracktablemodel.cpp:
```diff
@@ -762,6 +774,23 @@ QVariant BaseTrackTableModel::roleValue(
+                freq = int(Track::kDefaultTuningFrequency);
+            }
+            if (role == Qt::DisplayRole || role == Qt::ToolTipRole || role == kDataExportRole) {
+                return QStringLiteral("%1 Hz").arg(freq);
```
The unit can be moved to the header. Vertical space is rare. 
```suggestion
                return QStringLiteral("%1").arg(freq);
```

daschuer commented on src/library/basetracktablemodel.cpp:
```diff
@@ -762,6 +774,23 @@ QVariant BaseTrackTableModel::roleValue(
+                return QVariant();
+            }
+            if (freq == 0) {
+                freq = int(Track::kDefaultTuningFrequency);
```
It is probably better the return QVariant() here. The tuning is not known.

Commit:
fix: remove default frequency

Instead of showing some possibly wrong default frequency we show nothing.
This is the same behavior we have for key detection.

Commit:
fix: move pitch unit ("Hz") from cell into column header

We also give a more exact floating-point number in the tool-tip.

Commit:
fix: support tracks with mixed tuning frequencies

For example a mixed track might be made up of tracks using different
tuning frequencies.

Commit:
fix: shift tuning key for cent values >50 or <-50

quaklet commented:
> My eye opener for the whole Harmonic mixing myth was this: https://youtu.be/tCsl6ZcY9ag?si=yzZYbC8Zx-WCLGHK 

Very interesting, thank you! I had a rough idea what this was about, but the video helped to understand it much better. :)

> @danferns hat the idea to do a "heat" notation for this. ↑ ↓ an be replaced with a changing key coloring like we have in the key widget. @danferns has offered to contribute it.
> For now I have no strong opinion if we should keep them. Currently helpful but soon replaced ... Decide from you preferences.
> [...]
> It also depends if you plan to rewrite the history to eliminate ✧. This is not required. If no,t it does not matter when it is removed.

Sounds good!
I have no plans to rewrite the history here.
I'll just leave everything as it is then, to avoid complicating things.

daschuer reviewed (approved):
LGTM. I am very happy about the team work making this happen.
